version test - improve message when value is empty

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
When an empty value is provided, no `version` attribute will exist on the `LooseVersion` or `StrictVersion` object. We catch and handle this, but it's not immediatebly clear to the non-Python programmer that an `AttributeError` means an empty value was provided to the test.

Handle this case specifically and add a more helpful error message.

Add unit and integration tests.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/plugins/test/core.py`

samdoran committed:
version test - improve message when value is empty

When an empty value is provided, no `version` attribute will exist on the `LooseVersion` or
`StrictVersion` object. We catch and handle this, but it's not immediatebly clear that an
AttributeError means an empty value was provided.

Handle this case specifically and add a more helpful error message.

Add unit and integration tests.

sivel commented on lib/ansible/plugins/test/core.py:
```diff
@@ -168,6 +168,9 @@ def version_compare(value, version, operator='eq', strict=None, version_type=Non
     if strict is not None and version_type is not None:
         raise errors.AnsibleFilterError("Cannot specify both 'strict' and 'version_type'")
 
+    if not value:
```
I think we need to check both `version` and `value`.  Otherwise just catch the `AttributeError` in the below try/except and give this error there.

mattclay replied:
Both should be checked.

Do we want one error message, or a separate error message for each combination (value, version, both)?

samdoran replied:
Catching the `AttributeError` was the first approach I took. Then I switched to failing early by checking up front. I didn't think about checking both `value` and `version`, though.

I'll just test both here and adjust the error message.

samdoran replied:
I went with separate error messages for both. The language was too ambiguous trying to cram it into one message.

samdoran replied:
I'm open to feedback on the language. It's tough trying to communicate the _input_ version value versus the version comparison value. They're both version comparison values. 🤯

mattclay commented on test/units/plugins/filter/core/test_version_compare.py:
```diff
@@ -0,0 +1,17 @@
+@pytest.mark.parametrize('value', ('', None, set(), tuple(), dict(), 0, False))
+def test_version_compare_empty_value(value):
+    with pytest.raises(AnsibleFilterError, match='Version to compare cannot be empty'):
+        version_compare(value, '1.0')
```
I'd skip the unit test and just check this as part of the assert in the integration test instead.

samdoran replied:
But I like unit tests. 😜

samdoran committed:
Remove unit tests

samdoran committed:
Test both value and version

Add different error messages for each. Not handling the scenario if both are empty for now.
Update integration tests.
